[RFC] TS migration patterns

jbalsas opened the pull request:
Hey @liferay-frontend!! 

This is an initial exploration of what a migration to TS would look like for some of our modules and how ready our infrastructure is for other teams to safely adopt TS in their own modules.

I'm sharing so we can discuss about the patterns and possible impediments we have to fully transition to TS and make sure our infra is solid for other teams to move onto it soon.

Each commit message has some information about specific topics that might be worth discussing, and I'll be highlighting some things in particular, but feel free to review this at your will and raise any questions you might think of.

@bryceosterhaus, @matuzalemsteles, @izaera, mentioning you in particular since you might have the most experience in TS codebases... did we ever codify some guidelines or best practices in Clay or the other repos that we used to test TS with? Any hints to share? Thoughts? 👼

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-28.liferay.com/job/test-portal-source-format/575/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-28.liferay.com//userContent/jobs/test-portal-source-format/builds/575/jenkins-report.html">here</a>.</b></p>

jbalsas commented on modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/init.ts:
```diff
@@ -130,8 +130,24 @@ export default function init(config) {
+// Properties from `frontend-js-spa-web/init`
+declare global {
+	export interface ILiferay {
+		SPA?: {
```
Moved here, this could simply be `SPA?: typeof App`

matuzalemsteles replied:
Yeah, I think it would be better because eventually we can modify the App and forget to update the types here.

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>relevant</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-28.liferay.com/job/test-portal-acceptance-pullrequest(master)/669/">test-portal-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-28.liferay.com//userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/669/jenkins-report.html">here</a>.</b></p>

jbalsas commented on modules/apps/frontend-js/frontend-js-state-web/src/main/resources/META-INF/resources/State.ts:
```diff
@@ -16,6 +16,7 @@ import SubscriberMap from './SubscriberMap';
 import deepFreeze from './deepFreeze';
 
 import type {Immutable} from './types';
+import * as Types from 'frontend-js-web/src/main/resources/META-INF/resources/liferay/types';
```
Not loving this 😂

izaera replied:
:heart_eyes:

matuzalemsteles replied:
It would be great to add global types in a custom folder to be added via `typeRoots` to avoid importing types explicitly.

matuzalemsteles replied:
Hmm to some extent I agree with that but it's a little weird because you wouldn't be using `Liferay` directly because they are declared globally, that's why I find it weirder because normally this behavior is so that it's visible to all files because we are adding to `window`, opposed to defining the specific types of a module. For example, it might make sense if it had a syntax like:

```typescript
import type '/path/types';
```

Maybe if the types for Liferay weren't `declare global` but that wouldn't work. Either way it would be the same thing as the global types for Window.

liferay-continuous-integration commented:
<html><h3>:x: ci:test:sf - 0 out of 1 jobs passed in 2 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/ce10eb3c2e2f3427b904c078ac8192ca5b7e8d37">ce10eb3c2e2f3427b904c078ac8192ca5b7e8d37</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/jbalsas/liferay-portal/tree/types">types</a><br/>Branch GIT ID: <a href="https://github.com/jbalsas/liferay-portal/commit/0e56c3174113dec4787dd2528d37c4038f4bec6f">0e56c3174113dec4787dd2528d37c4038f4bec6f</a></p>0 out of 1jobs PASSED<div><h4>1 Failed Jobs:</h4><ul><li><a href="https://test-1-28.liferay.com/job/test-portal-source-format/575/"><strike><strong>test-portal-source-format</strong></strike></a></li></ul></div><h5>For more details click <a href="https://test-1-28.liferay.com/userContent/jobs/test-portal-source-format/builds/575/jenkins-report.html">here</a>.</h5><pre><code>     [exec]   21:1  error  imports must be sorted by module name (expected: "@liferay/frontend-js-state-web/src/main/resources/META-INF/resources/State" &lt;&lt; "frontend-js-spa-web/src/main/resources/META-INF/resources/init")  @liferay/liferay/sort-imports
     [exec]   21:13  error  'SPATypes' is defined but never used. Allowed unused vars must match /^_/u.  @typescript-eslint/no-unused-vars
     [exec]   22:13  error  'StateTypes' is defined but never used. Allowed unused vars must match /^_/u.  @typescript-eslint/no-unused-vars
     [exec] 
     [exec] ??? 9 problems (9 errors, 0 warnings)
     [exec]   5 errors and 0 warnings potentially fixable with the `--fix` option.
     [exec] 
     [exec] 3 of 3 jobs failed
     [exec] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
     [exec] error Command failed with exit code 1.
     [exec] 
     [exec] &gt; Task :packageRunCheckFormat FAILED
     [exec] Gradle build finished at 2021-08-20 09:40:57.420.
     [exec] 3 actionable tasks: 2 executed, 1 up-to-date
     [exec] 
     [exec] See the profiling report at: file:///opt/dev/projects/github/liferay-portal/build/reports/profile/profile-2021-08-20-02-40-38.html
     [exec] A fine-grained performance profile is available: use the --scan option.
     [exec] 
     [exec] 
     [exec] FAILURE: Build failed with an exception.
     [exec] 
     [exec] * What went wrong:
     [exec] Execution failed for task ':packageRunCheckFormat'.
     [exec] &gt; Process 'command '/opt/dev/projects/github/liferay-portal/build/node/bin/node'' finished with non-zero exit value 1
     [exec] 
     [exec] * Try:
     [exec] Run with --info or --debug option to get more log output. Run with --scan to get full insights.
     [exec] 
     [exec] * Exception is:
     [exec] org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':packageRunCheckFormat'.
     [exec] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda\$executeIfValid\$1(ExecuteActionsTaskExecuter.java:208)
     [exec] 	at org.gradle.internal.Try\$Failure.ifSuccessfulOrElse(Try.java:263)
     [exec] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:206)
     [exec] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:187)</code></pre></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-28.liferay.com/job/test-portal-source-format/575/">test-portal-source-format#575</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-28.liferay.com/userContent/jobs/test-portal-source-format/builds/575/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/1379">liferay-frontend#1379</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1554711550">[master] ci:test:sf - jbalsas &gt; liferay-frontend - PR#1379 - 2021-08-20[02:38:51]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-28.liferay.com/job/publish-testray-report/1817/">publish-testray-report#1817</a><br/></div></div>

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/add_params.ts:
```diff
@@ -14,12 +14,8 @@
- * @review
  */
-export default function addParams(params, baseUrl) {
+export default function addParams(params: string | object, baseUrl: string) {
```
We most certainly don't want `object` here but something like `{ [key:string]: string }`... does that have a universal name?

Also... see comments about runtime type guards in the commits... keep'em or drop'em?

Additional future warning... `typeof params !== 'object'` might not be telling you what you thing is telling you... maybe we should lint about this or something, idk!

izaera replied:
> `{ [key:string]: string }`... does that have a universal name ?

Nope, but we can create one if it is very common... 

I usually create types for this kind of option arguments. So something like this I think it's possible:

```
addParamsOptions {
   knownOption1: string;
   knownOption2: boolean;
   [key:string]: string;
}
```

If there are no known options just the `[key:string]: string;` part. But having it as a named (and exported) type is usually handy when you need to pass those options through several function calls and have them typed.

Of course I'm not saying 100% of the time you must do this, but if it is a public API there are many benefits in having a named type for this, IMO.

matuzalemsteles replied:
If we don't have known properties on the object, instead of using `[key:string]: string;` we have the `Record` that was introduced in TypeScript, we can replace it with `Record<string, string>`;

jbalsas replied:
> we can replace it with Record<string, string>
> +1

Nice! Completely missed that! Works like a charm! ⭐ 

> I prefer we don't create a global [key:string]: string; helper, as I think it may lead people to abuse it when they don't want to define a proper interface.

This is likely to happen anyways, isn't it? They can even use `any` 😂 . But yeah, totally get your point, let's keep those to a minimum until we know what's really useful and valuable! 👍

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/build_fragment.ts:
```diff
@@ -12,12 +12,12 @@
 	div.innerHTML = `<br>${htmlString}`;
 
-	div.removeChild(div.firstChild);
+	div.removeChild(div.firstChild!);
```
Not sure if there's a better way... plaguing our codebase with non-null assertions doesn't feel right, yet it's certainly a powerful tool 😉

matuzalemsteles replied:
Yeah, I agree, as it is explicit above that we have a children I think this makes it safe

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/get_portlet_namespace.ts:
```diff
@@ -14,11 +14,8 @@
- * @review
  */
-export default function getPortletNamespace(portletId) {
+export default function getPortletNamespace(portletId: string) {
```
Technically, this takes `any`... so... runtime type guards, keep'em or drop'em? Should we be stricter with types like this, or better type as our current usage and then clean it up?

izaera replied:
> runtime type guards, keep'em or drop'em?

I would say you still need them in case someone calls you from plain JavaScript...

However, if we decide to do runtime type checks, we should create a small utility for that, so that code is cleaner and less error prone. Something like:

```
Ensure.isString(portletId);
```

That we put at the top of methods.

jbalsas replied:
> I would say you still need them in case someone calls you from plain JavaScript...

The thing that worries me here is that we're throwing `TypeError` exceptions on a bad call. It's not like we're gracefully putting the system in a place where it can recover, but rather crashing it to the ground at runtime. 

This case is particularly tricky because the main difference is that if a developer calls this like `getPortletNamespace(3)` it will get:

```
// With runtime type guards
Uncaught TypeError: portletId must be a string

// Without
_3_
```

At runtime, the uncaught error seems more harmful, although the silent valid return can let issues go unnoticed to the developer for a while.

The more common case I guess would be something like `addParams(3, 'http://www.liferay.com')`

```
// With runtime type guards
Uncaught TypeError: Parameter params must be an object or string

// Without
Uncaught TypeError: params.trim is not a function
```

Now, I'm not convinced one way or another, so just trying to keep the conversation going here. It's not like we have so many of these anyways, but maybe we could keep the runtime type checks for utilities that get exported globally (ie: accessible through `Liferay.foo`) but not for those that need you to `import {foo} from 'frontend-js-web`', assuming the latter will be more easily moved into TS

izaera replied:
I agree that runtime failures are nasty, but people can avoid them simply by using a linter or TS at build time. If they don't want to do that, then they are exposed to more frequent runtime failures, because that's one of the reasons why typed languages were invented :shrug: .

Also, not throwing is much worse because you are deferring the failure to -basically- a random point in the future or, even worse, let it fail silently in ways that are very hard to reason about.

Say the developer writes `addParams(3, 'http://www.liferay.com')` and, instead of failing with `Uncaught TypeError: params.trim is not a function`, the code simply ignores the second parameter (because it uses an `Object.keys` or something like that, instead of `trim`).

Then, the method invocation would completely ignore the `'http://www.liferay.com'` but the developer would be quite puzzled because all he sees is that he's passing the option (correctly, he thinks) but the method is not honoring it. How long will it take to realize what is the error?

And in production, this will lead to some kind of functional failure which users will suffer too, so, is that better than failing fast with an exception? I doubt it...

izaera replied:
Think about security too. If you use parameters as a type, but the developers provide other types, it can lead to subtle vulnerabilites (as with user input). In an API, any parameter should be considered like user input, IMO, since you don't have any control over it.

jbalsas replied:
> And in production, this will lead to some kind of functional failure which users will suffer too, so, is that better than failing fast with an exception? I doubt it...

If you're in production, you probably haven't failed fast enough, have you? 😝 

I get all the pros/cons of the different approaches, so I think the point I'm trying to make is... what should be our default and how consistent we want to be about it? Right now, we only do runtime type checks in a handful of places but most others don't have any guarantees. As an example, [`fetch`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/fetch.es.js), which is a widely used utility doesn't do any type checking.

Should we:
- Drop all runtime type checks for typed APIs
- Keep (and add new) runtime type checks based on some rule of thumb. If so, which APIs should have checks and which ones shouldn't?
- Add runtime type checks everywhere

(I know this is a bit of a reductio ad absurdum because we probably don't want to do the last thing... or do we? 😉 )

bryceosterhaus replied:
>  With TS we can still get runtime errors

_[Elm](https://elm-lang.org/) has entered the chat_

julien replied:
:joy:

izaera replied:
>  I really don't want to make a distinction between people who choose TS and those who choose JS

It's not a distinction between TS and JS but between having a build with type checking or no build at all. Wether that type checking build is implemented with TS, eslint, or any other tool is an implementation detail.

Regarding runtime checks, I'm a big fan of super-robust public APIs for typed languages as well as dynamic ones, so I would make them (and probably design some way to make them automatic, not hand made), as I said above :shrug:

julien replied:
> It's not a distinction between TS and JS

I was referring to this part :wink:

> ... people using plain JS are on their own (and should suffer, at least a bit, for choosing not to use types).

But, I understand what you said about runtime checks

carloslancha replied:
> (and should suffer, at least a bit, for choosing not to use types)

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia1.giphy.com%2Fmedia%2FdeaabWAgiawzyiMNZS%2Fgiphy.gif&token=kill" data-gitmeme-token="kill"><img src="https://media1.giphy.com/media/deaabWAgiawzyiMNZS/giphy.gif" title="Created by gitme.me with /kill"/></a>

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/navigate.ts:
```diff
@@ -16,34 +16,32 @@
+	url: string | URL,
+	listeners?: {[key: string]: Callback}
+) {
+	url = url instanceof URL ? String(url) : url;
```
We do a lot of type narrowing that TS can't pick up like in here. In this case, `instanceof` sounds like a better approach overall, but wondering what we do in general.

We're likely going to need to rewrite code because this code style is not easily expressed with types (that I could figure). There are other ways to guard types like user-defined or creating intermediate helper methods like `getURLString(url: string | URL): string` that help TS here. 

Wondering how this will shape out in some of the more complex isomorphisms we have 😈

izaera replied:
I have a similar in [FilePath](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/js-toolkit/packages/js-toolkit-core/src/file/FilePath.ts) I think...

I use the [AnyPath](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/js-toolkit/packages/js-toolkit-core/src/file/FilePath.ts#L9) type for `string | FilePath` and [FilePath.coerce](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/js-toolkit/packages/js-toolkit-core/src/file/FilePath.ts#L14) to do the cast.

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_portlet_url.ts:
```diff
@@ -95,7 +94,7 @@ export default function createPortletURL(basePortletURL, parameters = {}) {
 
 	if (Object.entries(parameters).length) {
-		namespace = getPortletNamespace(portletID);
+		namespace = getPortletNamespace(portletID!);
```
For sure `portletID` can be `null` here. That's a possible bug somewhere that we haven't yet seen. Thoughts on what should we really do here?

markocikos replied:
Few lines above:
```js
if (Object.entries(parameters).length && !portletID) {
	throw new TypeError(
		'Portlet ID must not be null if parameters are provided'
	);
}
```

markocikos replied:
Generally speaking, I guess we could replace all manual type checks with TS. Unless we want them in runtime, to show some legitimate case in an error log :thinking:

jbalsas replied:
So that should be a `||` 😂

jbalsas replied:
You see, the thing is that if parameters aren't provided, then `portletID` **can** be null, but then `getPortletNamespace` takes (or could take) non-nullable strings. That's the thing I wanted to raise the attention to. Need to decide how much we lean on the type system. I think the more the better, but maybe in this case `null` is a valid thing to pass to `getPortletNamespace` which doesn't seem to be the case since it will throw a `TypeError` in that case.

markocikos replied:
If parameters are not provided, we never reach `getPortletNamespace`:
```js
if (Object.entries(parameters).length) {
	namespace = getPortletNamespace(portletID);
}
```

With this, the value of `portletID`, that we took it from `basePortletURL`, does not matter.

The bug here might be that we are not checking and namespacing params in `basePortletURL`. But, I'm guessing we are assuming the `basePortletURL` will be used as intended, a base URL, and not as any URL that we are appending out params on.

jbalsas replied:
> If parameters are not provided, we never reach getPortletNamespace:

lol, you're totally right... need to get me some better reading glasses 👴 

I wonder why TS is unable to infer that `portletID` can't be `null` at that point. Maybe this could be written in a way that it would make it possible similar to what happens in https://github.com/liferay-frontend/liferay-portal/pull/1379#discussion_r692819269

For this case I'll simply `as string`, then.

izaera replied:
Maybe putting everything inside one `if`...

```javascript
let namespace = '';

if (Object.entries(parameters).length) {
    if (!portletID) {
        throw new TypeError(
            'Portlet ID must not be null if parameters are provided'
        );
    }
    namespace = getPortletNamespace(portletID);
}
```

jbalsas replied:
Yeah, that should do it!

jbalsas commented on modules/apps/frontend-js/frontend-js-web/test/liferay/util/add_params.ts:
```diff
@@ -65,9 +65,7 @@ describe('Liferay.Util.addParams', () => {
-		expect(() => addParams(1)).toThrow(TypeError);
-
-		expect(() => addParams(['one', 'two'])).toThrow(TypeError);
+		expect(() => addParams(1 as any, sampleUrl)).toThrow(TypeError);
```
Another annoyance of keeping runtime type guards around... in tests, to test them you need `foo as any` :)

Funny enough, the second test was passing because the implementation throws `TypeError` from 2 different code branches, not because `typeof [] !== 'object'` 😉

jbalsas replied:
I think that if we keep the runtime assertions, then we'd likely want to keep the tests as well?

https://github.com/liferay-frontend/liferay-portal/pull/1379#discussion_r694257114

Following the same reasoning, there's no guarantee people will call this from TS so chances are we won't be testing types?

jbalsas commented on modules/apps/frontend-js/frontend-js-web/test/liferay/util/portlet_url/create_action_url.ts:
```diff
@@ -12,17 +12,15 @@
 				getPortalURL: jest.fn(() => 'http://localhost:8080'),
 			},
-		};
+		} as any;
```
Mocking can get tricky... 🤔

izaera replied:
Yes. But that's expected. That's why you have to use libraries like mockito in Java, too :shrug:

matuzalemsteles replied:
Maybe using `as const` can be more secure, at least we can have the viability of what we have inside the Liferay mock.

jbalsas commented:
ci:stop

liferay-continuous-integration commented:
<html><h3>:x: ci:test:relevant - 0 out of 1 jobs passed in 15 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/ce10eb3c2e2f3427b904c078ac8192ca5b7e8d37">ce10eb3c2e2f3427b904c078ac8192ca5b7e8d37</a></p><h4>Upstream Comparison:</h4><p>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/ce10eb3c2e2f3427b904c078ac8192ca5b7e8d37">ce10eb3c2e2f3427b904c078ac8192ca5b7e8d37</a><br/>Jenkins Build URL: <a href="https://test-1-1.liferay.com/job/test-portal-acceptance-upstream-dxp(master)//2232">Acceptance Upstream DXP (master) #2232</a></p><details><summary><strong>ci:test:relevant - 0 out of 1 jobs PASSED</strong></summary><div><h4>1 Failed Jobs:</h4><ul><li><a href="https://test-1-28.liferay.com/job/test-portal-acceptance-pullrequest(master)/669/"><strike><strong>test-portal-acceptance-pullrequest(master)</strong></strike></a></li></ul></div></details><h5>For more details click <a href="https://test-1-28.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/669/jenkins-report.html">here</a>.</h5><hr/><h4>Failures unique to this pull:</h4><ol><li><div><h5><a href="https://test-1-28.liferay.com/job/test-portal-acceptance-pullrequest(master)/669/">test-portal-acceptance-pullrequest(master)</a></h5><div><h6>Job Results:</h6><p>0 Jobs Passed.<br/>1 Job Failed.</p></div><pre><code>Build was aborted</code></pre></div></li></ol><h4>For upstream results, click <a href="https://test-1-1.liferay.com/job/test-portal-acceptance-upstream-dxp(master)">here</a>.</h4></details></html><h5>Test bundle downloads:</h5><ul><li><a href="https://files.liferay.com/private/bundles/test-1-28/jobs/test-portal-acceptance-pullrequest(master)/builds/669/liferay-portal-bundle-tomcat.tar.gz">liferay-portal-bundle-tomcat.tar.gz</a> (mirrors: <a href="http://mirrors.dlc.liferay.com/files.liferay.com/private/bundles/test-1-28/jobs/test-portal-acceptance-pullrequest(master)/builds/669/liferay-portal-bundle-tomcat.tar.gz">dlc</a>, <a href="http://mirrors.lax.liferay.com/files.liferay.com/private/bundles/test-1-28/jobs/test-portal-acceptance-pullrequest(master)/builds/669/liferay-portal-bundle-tomcat.tar.gz">lax</a>)</li></ul>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-28.liferay.com/job/test-portal-acceptance-pullrequest(master)/669/">test-portal-acceptance-pullrequest(master)#669</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-28.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/669/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>relevant<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/1379">liferay-frontend#1379</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1554712172">[master] ci:test:relevant - jbalsas &gt; liferay-frontend - PR#1379 - 2021-08-20[02:38:48]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-28.liferay.com/job/publish-testray-report/1819/">publish-testray-report#1819</a><br/></div></div>

izaera commented on modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx:
```diff
@@ -47,10 +47,10 @@ export default function render(
 	container: Element
 ) {
-	if (!(window.Liferay as any).SPA || (window.Liferay as any).SPA.app) {
+	if (!Liferay.SPA || Liferay.SPA.app) {
```
We should definitely provide typings for `Liferay` so that people can use it as documentation :+1: 

When I did the sample Angular project for the JS Toolkit, I had to use `declare Liferay: any` because we were lacking that, but it was clear that publishing the types for our `Liferay` interface would be a win (at least for the public parts).

izaera commented on modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/init.ts:
```diff
@@ -17,6 +17,8 @@ import ActionURLScreen from './screen/ActionURLScreen';
 import RenderURLScreen from './screen/RenderURLScreen';
 import {getUrlPath} from './util/utils';
 
+import * as Types from 'frontend-js-web/src/main/resources/META-INF/resources/liferay/types';
```
:heart_eyes:

izaera commented on modules/apps/frontend-js/frontend-js-state-web/src/main/resources/META-INF/resources/State.ts:
```diff
@@ -514,11 +515,8 @@ window.Liferay.State = State;
-			State: typeof State;
-		};
+	export interface ILiferay {
+		State?: typeof State;
```
How does this work? Is it merged with `SPA` in https://github.com/liferay-frontend/liferay-portal/pull/1379/files#diff-10174dbf5a248993084321a6662596ff7faa394aed5696d0fd001030e3544a0bR140 ?

jbalsas replied:
Yeah, it'll get merged with whatever global declaration of the interface is there. The thing is you need to make sure everyone contributing those types is loaded all the time which is not the case currently... need to figure out a different approach, likely involving `@types` and `typeRoots` or sth like that... 🤔

jbalsas replied:
Of course, even better would be if we stopped using the `Liferay` global object 😂

izaera replied:
Of course, of course,... :sweat_smile:

markocikos commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/index.ts:
```diff
@@ -16,88 +16,178 @@
 export {default as navigate} from './liferay/util/navigate.es';
+
+// @ts-ignore
+
```
Is there a way to `@ts-ignore` the whole file?

jbalsas replied:
Yes, you can do `// @ts-nocheck`, but then I wouldn't be able to slowly clean them up 😉

markocikos commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/types.ts:
```diff
@@ -132,6 +132,32 @@ export interface ILiferay {
+
+// Properties from `frontend-js-spa-web/init`
+
+export interface ILiferay {
```
Will the second `export interface ILiferay` override the first?

jbalsas replied:
No, they merge as per [Declaration Merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) which is what I thought we could leverage to compose them from different modules

markocikos replied:
Interesting! Looks powerful, but also quite quirky, with a lot of edge cases.

bryceosterhaus commented:
> did we ever codify some guidelines or best practices in Clay or the other repos that we used to test TS with? Any hints to share? 

Not sure what you mean by testing TS with. We do have a short [TS Guidlines doc](https://github.com/liferay/liferay-frontend-projects/blob/master/guidelines/general/typescript.md) and especially with react we recommend [react-typescript-cheatsheet](https://github.com/typescript-cheatsheets/react). Honestly though, I think the most helpful thing is to check in other places how TS is used, it is one of those tools that can be really overdone and you spend 95% of your time creating complex type structures. Generally I like to aim for simple and concise. It is going to take an external use a significant amount of time just to understand the type, I think we should go simpler.

jbalsas commented:
Hey @bryceosterhaus!

> Not sure what you mean by testing TS with

Pardon my english 😂 . By testing I meant "try out". We rolled it out in some of those projects so that we could get some hands-on experience and some initial view on possible issues and mitigations to help others get on board easily. So, my question is... what were those learnings? Which things went well, what helped, what didn't...? 

Hope that makes it clear-er :)

bryceosterhaus reviewed (commented):
Overall, this looks pretty good to me! It doesn't look nearly as bad as I might have originally thought it would.

> So, my question is... what were those learnings? Which things went well, what helped, what didn't...?

I don't think there were any learnings that were concrete or "one-size-fits-all" enough to document.

Personally, at a high-level view I have made a few observations about TS
- Migrating existing code to TS is usually harder than I originally think
- Writing TS code as a whole is typically easier than I think
- The more TS you have, the easier it is to write and migrate TS.
- Copy TS patterns from existing code (both your own and 3rd party)
- Always aim for simple types, type gymnastics typically indicate a code smell

matuzalemsteles commented on modules/apps/frontend-js/frontend-js-web/test/liferay/util/portlet_url/create_portlet_url.ts:
```diff
@@ -12,34 +12,35 @@
 describe('Liferay.Util.PortletURL.createPortletURL', () => {
 	afterEach(() => {
-		Liferay.ThemeDisplay.getPortalURL.mockRestore();
+		(Liferay.ThemeDisplay.getPortalURL as jest.Mock<
+			string,
+			[]
+		>).mockRestore();
```
Maybe to avoid things like this for global objects we can create a base interface for Liferay with generic types to be able to use it in tests and another one for development.

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.ts:
```diff
@@ -12,7 +12,7 @@
  */
 
-const MAP_HTML_CHARS_ESCAPED = {
+const MAP_HTML_CHARS_ESCAPED: {[key: string]: string} = {
```
True that! Thanks! 👍

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.ts:
```diff
@@ -24,7 +24,7 @@ const MAP_HTML_CHARS_ESCAPED = {
 export {MAP_HTML_CHARS_ESCAPED};
 
-const MAP_HTML_CHARS_UNESCAPED = {};
+const MAP_HTML_CHARS_UNESCAPED: {[key: string]: string} = {};
```
Similar to https://github.com/liferay-frontend/liferay-portal/pull/1379#discussion_r694536455 and on top of that, we also do:

```javascript
Object.entries(MAP_HTML_CHARS_ESCAPED).forEach(([char, escapedChar]) => {
	MAP_HTML_CHARS_UNESCAPED[escapedChar] = char;
});
```

Which also requires string-based key access. I get the feeling this could be reworked somehow if we wanted, but should be okay for now...

jbalsas commented:
Hey @izaera, @markocikos, @bryceosterhaus, @matuzalemsteles, I've pushed some additional conversions in case you want to take a look.

I removed the global `Liferay` ambient namespace approach to reduce the noise and so far I haven't been able to find a nice way for different modules to easily collaborate to a global namespace (which is a rather bad practice anyways). I haven't been able to reliable find a way get this done. So far, the reliable options I have are:
- Put everything `Liferay` inside `frontend-js-web` since most other projects will depend on it and see the types
- Use [triple slash path directives](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html#-reference-path-). This works, but one needs to specify the paths to every piece of the declaration through several modules
- Use type (or any other) type of imports to bring in the types. Same as with path directives, need to import as many as you want. Keeping in mind that if you only care about `Liferay.SPA` you'd only really need to import types from `frontend-js-spa-web` maybe makes this bearable. The only drawback is that you need to know where the global part of `Liferay` is coming from.
- Create some type of intermediate project where we can do the aggregation of Liferay bits so we only need to import that one (and additionally maybe publish it so we can get it from `@types`)

From time to time, other things like project references or typeRoots seem to work, but then they stop working, so I'm not 100% sure yet what's going on.

I'd say, following @bryceosterhaus's advice that we could start small by putting everying inside the same `liferay.d.ts` in `frontend-js-web` and live with it and hopefully reduce the amount of global access we have overtime. We can also split it up and load the types on a per-usage need 🤷  Thoughts?

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/components/SimpleInputModal.es.js:
```diff
@@ -61,6 +61,7 @@ const SimpleInputModal = ({
 			document.querySelector(`#${namespace}form`)
 		);
 
+		// eslint-disable-next-line @liferay/portal/no-global-fetch
```
@bryceosterhaus, @kresimir-coko, this will need an adjustment to `no-global-fetch` since we explicitly look for `fetch.es` as a valid import which we won't have any longer... do you think we could work on this while this PR sits over here?

kresimir-coko replied:
It's definitely something that we can investigate, I'll open an issue for it.

kresimir-coko replied:
Awesome!

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/align.ts:
```diff
@@ -12,52 +12,56 @@
+		? node.documentElement
+		: node;
+
+	return element[`client${property}` as ClientProp];
```
Brackets access is always tricky... 🤔

jbalsas commented on modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/is_object.ts:
```diff
@@ -16,7 +16,7 @@
  * and functions.
  */
-export default function isObject(val: any) {
+export default function isObject<T>(val: T): val is NonNullable<T> {
```
I think this should infer `never` when called with other types. For example, `const a = 'a'; if (isObject('a')) {...}` will consider that a is of type `string` instead of `never`.

jbalsas commented:
Hey @bryceosterhaus, @matuzalemsteles, I found some time today to try something out. I was inspired by [this example of usage from prisma](https://github.com/prisma/prisma-examples/blob/latest/typescript/rest-nextjs-api-routes/tsconfig.json#L25).

So, the idea is that we could concentrate our global definition indirections in [`/modules/global-env.d.ts`](https://github.com/liferay-frontend/liferay-portal/pull/1379/commits/91c8d47c26807f097a80af7920c3dcded6464fd4#diff-f8db832734387a5f3b1449bb1c983ebc3c7ccb0284e8ac570d7c13324011c711) and from there point to the different modules that offer globals:
- [frontend-js-web::global.d.ts](https://github.com/liferay-frontend/liferay-portal/pull/1379/commits/91c8d47c26807f097a80af7920c3dcded6464fd4#diff-090e8f4a219606c9bf34ac6fd52580571f54dabf8ef4aaca091ba15583a9ab64)
- [frontend-js-spa-web::global.d.ts](https://github.com/liferay-frontend/liferay-portal/pull/1379/commits/91c8d47c26807f097a80af7920c3dcded6464fd4#diff-523f727d06b6d445a9035cffc6acf34100cac07d9d973cfaaff267e3e60fab95)
- [frontend-js-state-web::global.d.ts](https://github.com/liferay-frontend/liferay-portal/pull/1379/commits/91c8d47c26807f097a80af7920c3dcded6464fd4#diff-236b6f3312b1dabf7a59524e6b6e4df3f2d9dec8bb31f6fe963498cada754d61)

To wrap this up, we'd simply need to add a resolution to the root file in [configureTypeScript](https://github.com/liferay/liferay-frontend-projects/blob/master/projects/npm-tools/packages/npm-scripts/src/typescript/configureTypeScript.js#L100-L124) so that we don't need to add overrides.

What do you think, does this sound like a sensible alternative to unblock TS usage in Portal? If so, I can send the change to the `npm-scripts` project and then clean this up to give it a try. I think there's enough work done here already to grant an attempt.

/cc @nhpatt

bryceosterhaus commented:
> What do you think, does this sound like a sensible alternative to unblock TS usage in Portal?

Yeah, this sounds like a reasonable plan to me and isn't so risky that it could dramatically impact us negatively.

As always, I always like the idea of trying out something simple and seeing how it goes. Especially since this is a dev related, it makes it easier for us to iterate.

matuzalemsteles commented:
Oh I really like it too, it feels simpler and cleaner compared to having to import it, my 👍 to go with it 😁.

jbalsas commented:
Well... I think it's obvious that I won't get live to see this done... 😢 

I just rebased and pushed the branch with what I had so far. I hit a bit of an egg and chicken type of problem with the current approach of a `global.env.d.ts` file since this makes some modules (like `frontend-js-web`) fail to compile because it finds sources outside `rootDir`. I tried several different approaches but wasn't really successful, so fixing this is now up to you! 💪 

I'll keep this repo and branch around for some time, but it would be advisable for someone to keep track of it independently just in case I get trigger-happy with the "delete repo" button 😂 

![amazingjourney](https://user-images.githubusercontent.com/905006/142493237-e9d98a00-5c27-42f2-b5fb-72cfcf592551.gif)

bryceosterhaus commented:
> Well... I think it's obvious that I won't be around long enough to see this done... 😢

We actually have been meaning to tell you, you aren't allowed to leave until this is done... Glad to have you around awhile! 😂

kresimir-coko commented:
I pulled this, rebased and ran `gradlew clean deploy` from `frontend-js-web`, and here's the [gist of all the errors it produced](https://gist.github.com/kresimir-coko/5b2b5dd308a88499c436d695480a0c60).

Most of the errors seem to be coming from `frontend-js-spa-web` and `frontend-js-state-web`, with a few of them coming from  a couple files inside `test/liferay/util/...`

The commit containing changes triggered by `formatSource`  and `deploy` can be found on https://github.com/kresimir-coko/liferay-portal/commit/b306dd8c7e655da6928279df4d762ed037014e1b

@bryceosterhaus I guess next steps would be to get rid of these errors so that it can be actually deployed

julien commented:
> @bryceosterhaus I guess next steps would be to get rid of these errors so that it can be actually deployed

I guess you're right :+1:

kresimir-coko commented:
Closed in favor of #1732
